regression in "Build failed with error TS2416 in firebase-functions"

I'm filing this because there's been a serious regression on an old issue and there's a lot of discussion/new comments that aren't being responded to, I believe due to the issue being closed. Can we get some eyes on it?

Issue that's regressed: https://github.com/firebase/firebase-functions/issues/1442

I am having the same issue when compiling my TS code to JS with these packages installed:
```
    "firebase-admin": "^12.0.0",
    "firebase-functions": "^4.6.0",
```

Output:
```
$ npm run clean && tsc

node_modules/firebase-functions/lib/common/providers/database.d.ts:80:5 - error TS2416: Property 'child' in type 'DataSnapshot' is not assignable to the same property in base type 'DataSnapshot'.
  Type '(childPath: string) => DataSnapshot' is not assignable to type '(path: string) => DataSnapshot'.
    Call signature return types 'DataSnapshot' and 'DataSnapshot' are incompatible.
      The types of 'forEach' are incompatible between these types.
        Type '(action: (a: DataSnapshot) => boolean | void) => boolean' is not assignable to type '(action: (a: IteratedDataSnapshot) => boolean | void) => boolean'.
          Types of parameters 'action' and 'action' are incompatible.
            Types of parameters 'a' and 'a' are incompatible.
              Type 'DataSnapshot' is not assignable to type 'IteratedDataSnapshot'.
                Types of property 'key' are incompatible.
                  Type 'string | null' is not assignable to type 'string'.
                    Type 'null' is not assignable to type 'string'.
```

I have the exact same issue.

> I am having the same issue when compiling my TS code to JS with these packages installed:
> 
> ```
>     "firebase-admin": "^12.0.0",
>     "firebase-functions": "^4.6.0",
> ```
> 
> Output:
> 
> ```
> $ npm run clean && tsc
> 
> node_modules/firebase-functions/lib/common/providers/database.d.ts:80:5 - error TS2416: Property 'child' in type 'DataSnapshot' is not assignable to the same property in base type 'DataSnapshot'.
>   Type '(childPath: string) => DataSnapshot' is not assignable to type '(path: string) => DataSnapshot'.
>     Call signature return types 'DataSnapshot' and 'DataSnapshot' are incompatible.
>       The types of 'forEach' are incompatible between these types.
>         Type '(action: (a: DataSnapshot) => boolean | void) => boolean' is not assignable to type '(action: (a: IteratedDataSnapshot) => boolean | void) => boolean'.
>           Types of parameters 'action' and 'action' are incompatible.
>             Types of parameters 'a' and 'a' are incompatible.
>               Type 'DataSnapshot' is not assignable to type 'IteratedDataSnapshot'.
>                 Types of property 'key' are incompatible.
>                   Type 'string | null' is not assignable to type 'string'.
>                     Type 'null' is not assignable to type 'string'.
> ```

I am getting this error right now with: 

```
  "dependencies": {
    "@faker-js/faker": "^8.4.0",
    "@types/google-libphonenumber": "^7.4.30",
    "@types/nodemailer": "^6.4.14",
    "axios": "^1.6.7",
    "firebase-admin": "^12.0.0",
    "firebase-functions": "^4.6.0",
    "fp-ts": "^2.16.2",
    "google-libphonenumber": "^3.2.34",
    "meilisearch": "^0.37.0",
    "nodemailer": "^6.9.8",
    "urlcat": "^3.1.0",
    "uuid": "^9.0.1"
  },
```

The error I get is:

```
npm run build

> build
> tsc

node_modules/firebase-functions/lib/common/providers/database.d.ts:80:5 - error TS2416: Property 'child' in type 'DataSnapshot' is not assignable to the same property in base type 'DataSnapshot'.
  Type '(childPath: string) => DataSnapshot' is not assignable to type '(path: string) => DataSnapshot'.
    Call signature return types 'DataSnapshot' and 'DataSnapshot' are incompatible.
      The types of 'forEach' are incompatible between these types.
        Type '(action: (a: DataSnapshot) => boolean | void) => boolean' is not assignable to type '(action: (a: IteratedDataSnapshot) => boolean | void) => boolean'.
          Types of parameters 'action' and 'action' are incompatible.
            Types of parameters 'a' and 'a' are incompatible.
              Type 'DataSnapshot' is not assignable to type 'IteratedDataSnapshot'.
                Types of property 'key' are incompatible.
                  Type 'string | null' is not assignable to type 'string'.
                    Type 'null' is not assignable to type 'string'.

80     child(childPath: string): DataSnapshot;
       ~~~~~

node_modules/firebase-functions/lib/common/providers/database.d.ts:100:5 - error TS2416: Property 'forEach' in type 'DataSnapshot' is not assignable to the same property in base type 'DataSnapshot'.
  Type '(action: (a: DataSnapshot) => boolean | void) => boolean' is not assignable to type '(action: (a: IteratedDataSnapshot) => boolean | void) => boolean'.

100     forEach(action: (a: DataSnapshot) => boolean | void): boolean;
        ~~~~~~~


Found 2 errors in the same file, starting at: node_modules/firebase-functions/lib/common/providers/database.d.ts:80
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -1,1 +1,2 @@
 - Fixes access on deeply nested, nonexistent property. (#1432)
+- Add IteratedDataSnapshot interface to match with firebase admin v12 (#1517).

--- a/src/common/providers/database.ts
+++ b/src/common/providers/database.ts
@@ -26,7 +26,7 @@ import { firebaseConfig } from "../../common/config";
 import { joinPath, pathParts } from "../../common/utilities/path";
 
 /**
- * Pulled from @firebase/database-types
+ * Pulled from @firebase/database-types, make sure the interface is updated on dependencies upgrades.
  * Represents a child snapshot of a `Reference` that is being iterated over. The key will never be undefined.
  */
 interface IteratedDataSnapshot extends DataSnapshot {

--- a/src/common/providers/database.ts
+++ b/src/common/providers/database.ts
@@ -26,6 +26,14 @@ import { firebaseConfig } from "../../common/config";
 import { joinPath, pathParts } from "../../common/utilities/path";
 
 /**
+ * Pulled from @firebase/database-types
+ * Represents a child snapshot of a `Reference` that is being iterated over. The key will never be undefined.
+ */
+interface IteratedDataSnapshot extends DataSnapshot {
+  key: string; // key of the location of this snapshot.
+}
+
+/**
  * Interface representing a Firebase Realtime database data snapshot.
  */
 export class DataSnapshot implements database.DataSnapshot {
@@ -204,7 +212,7 @@ export class DataSnapshot implements database.DataSnapshot {
    * @return `true` if enumeration was canceled due to your callback
    *   returning `true`.
    */
-  forEach(action: (a: DataSnapshot) => boolean | void): boolean {
+  forEach(action: (a: IteratedDataSnapshot) => boolean | void): boolean {
     const val = this.val() || {};
     if (typeof val === "object") {
       return Object.keys(val).some((key) => action(this.child(key)) === true);

